### ISSUE TITLE
'createUser' RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by merlinpatt** [meteor/meteor#8510]
Versions: 1.4.3.2, 1.4.2.3
Platform: Ubuntu 14.04

I was having this issue in 1.4.2.3 and so I upgraded to 1.4.3.2 in hopes that it would fix this, but it hasn't. I also attempted `meteor reset` but that hasn't worked either.

I am unable to create a user using `createUser` and `loginWithGoogle`. 

I am calling `createUser` with `Accounts.createUser({username, password, profile, role: 'student'})`. And I am calling `loginWithGoogle` with no arguments. Both give the same error trace below.

```
Exception while invoking method 'createUser' RangeError: Maximum call stack size exceeded
     at Object.EJSON.clone (packages/ejson.js:482:24)
     at Object.EJSON.clone (packages/ejson.js:507:22)
     at packages/ejson.js:521:22
     at Function._.each._.forEach (packages/underscore.js:147:22)
     at Object.EJSON.clone (packages/ejson.js:520:5)
     at packages/ejson.js:521:22
     at Function._.each._.forEach (packages/underscore.js:147:22)
     at Object.EJSON.clone (packages/ejson.js:520:5)
     at packages/ejson.js:521:22
     at Function._.each._.forEach (packages/underscore.js:147:22)
     at Object.EJSON.clone (packages/ejson.js:520:5)
```

**Comment by abernix** [comment]
It seems you're suggesting that `Accounts.createUser` has been broken since Meteor 1.4.2.3 which honestly seems to me like it would have thrown up a lot of flags in the last four months (since 1.4.2.3 was released) if it were true by itself.

Will you please provide a basic a reproduction repository per [reporting a bug](https://github.com/meteor/meteor/blob/devel/Contributing.md#reporting-bug) so that when someone goes to run your reproduction on Ubuntu 14.04 it's guaranteed to be an identical project? :smile:

**Comment by merlinpatt** [comment]
I'm certainly not trying to imply that it's been broken that long. Just listing the versions it has happened on to help find the issue more.

I'll work on a repro later today. In the meantime, is there a way for me to get further in the error trace?

If I can find out where `clone` is being called, I might be able to find the source of the error.

**Comment by merlinpatt** [comment]
I figured out the bug and feel dumb now. We're using simple schema, and one of the fields was specified as required but wasn't being passed in upon signup. I'm not sure why it wasn't reporting like it normally would though.

I am still curious about the loop and the internals causing that, so if you happen to know a good way to debug that, I'd love to know. I'd like to start digging more into Meteor's internals and help actively develop it.

Also, I seem to recall an issue or PR talking about removing underscore. Is that still happening / how can I help with that?

**Comment by abernix** [comment]
No worries.  Glad you got it figured out!  Max-stack traces can be difficult, but sometimes it's best to hop into `meteor debug` and see if it reveals anything obvious!

(Re: Underscore => Lodash, I think it's _basically_ done but there are a couple loose ends to tie up.  I've been busy with other pressing matters unfortunately.  Soon, hopefully, but follow #1009 for official updates.  And please, dig into Meteor internals and issues if you can!  See the new [`Development.md`](https://github.com/meteor/meteor/blob/devel/Development.md) from #8267 for starters!).

**Comment by harishreddy-m** [comment]
Similar error
User with existing account tried OAuth login button.OAuth redirection is successful.
Got internal server error on oauth client.

I20170323-13:06:39.122(5.5)? Exception while invoking method 'login' RangeError: Maximum call stack size exceeded
I20170323-13:06:39.123(5.5)?     at Object.EJSON.clone (packages/ejson/ejson.js:462:1)
I20170323-13:06:39.123(5.5)?     at packages/ejson/ejson.js:501:1
I20170323-13:06:39.123(5.5)?     at Function._.each._.forEach (packages/underscore/underscore.js:113:1)
I20170323-13:06:39.124(5.5)?     at Object.EJSON.clone (packages/ejson/ejson.js:500:1)
I20170323-13:06:39.124(5.5)?     at packages/ejson/ejson.js:501:1
I20170323-13:06:39.124(5.5)?     at Function._.each._.forEach (packages/underscore/underscore.js:113:1)
I20170323-13:06:39.124(5.5)?     at Object.EJSON.clone (packages/ejson/ejson.js:500:1)
I20170323-13:06:39.125(5.5)?     at packages/ejson/ejson.js:501:1
I20170323

Any quick fixes to unblock are appreciated.

P.S : I don't understand why this bug is closed.OP might have found the reason, but the error message is nowhere helping the developer.
Do you recommend posting this bug in accounts-oauth package?

**Comment by abernix** [comment]
@harishreddy-m If you can provide a simple reproduction it would make debugging and finding a solution more possible, but please open a new issue with that reproduction and reference this so we can look into it more.  The issue you've quoted seems to show a lot of looping (`forEach` within `forEach` within `forEach`, etc.) indicating that some sort of recursion is happening – hard to say without looking at the code/testing it.

**Comment by merlinpatt** [comment]
@harishreddy-m are you using SimpleSchema or any other schema packages?

**Comment by harishreddy-m** [comment]
I am using simple schema

**Comment by harishreddy-m** [comment]
Yes. I am using simple schema.

**Comment by merlinpatt** [comment]
@harishreddy-m remove all your schema and see if the app runs. If it does, the issue is with one of your schemas.

**Comment by harishreddy-m** [comment]
Removing schema is not valid solution for this problem.
On successful login for Oauth,
execution is entering `Ap.updateOrCreateUserFromExternalService` method 
```
 selector[serviceIdKey] = serviceData.id;
  }

  var user = this.users.findOne(selector);```

If it is first time login,it is trying to `userId = this.users.insert(fullUser);` 
This method is not making any attempt to merge existing user with same email.

This code is in accounts-base@1.2.11 package

**Comment by merlinpatt** [comment]
@harishreddy-m I'm not saying remove the schema permanently. I'm saying remove them temporarily to see if they are the cause.

**Comment by harishreddy-m** [comment]
Obviously, removing simple schema will remove the error.

loginWith<Service> takes a function to find user or/and populate user object would be good to have.

**Comment by merlinpatt** [comment]
If removing the schema, removes the error, then you probably have a field that is not being created when the user is created. Try making various fields optional, or give them a defaultValue. That's what we had to do

**Comment by harishreddy-m** [comment]
I made all fields in user schema as optional to get rid of this error.

Complete solution will be : accounts-oauth to let us inject fields into user document it tries to create.
Currently it creates user document with just _id,services,createdAt fields

**Comment by finderbar** [comment]
yesterday, i got this error after meteor update.
it's fine if i don't specific users schema.
i wanna to secure user login. i don't know how to sloved it.

![error](https://user-images.githubusercontent.com/24274777/32147316-b8570c02-bd13-11e7-8640-262bb1324e23.png)
